R/params.R

```diff
@@ -0,0 +1,329 @@
+          ## This is not current_value(param) because we want to understand
+          ## deviation from the report default, not the call-time override.
+          param$value
+        }
+      }
+      
+      inputControl <- do.call(inputControlFn, arguments)
+      useSelectControl <- function(current) { FALSE }
+      selectControl <- NULL
+
+      ## Dates and times with expressions that mean "now" or "today" are first
+      ## materialized as selects. If the user chooses to customize the field,
+      ## we then show the type-specific picker.
+      if (is.null(params[[param$name]])) { # prior value; implicit customization
+        if ("POSIXct" %in% param$class && identical("Sys.time()", param$expr)) {
+          useSelectControl <- function(current) {
+            (is.null(current) || identical(current, "default"))
+          }
+          uidefaults[[param$name]] <<- "default"
+          choices <- list()
+          choices[[paste0("now (",param$value,")")]] <- "default"
+          choices[["Use a custom time"]] <- "custom"
+          selectControl <- shiny::selectInput(inputId = inputId,
+                                              label = label,
+                                              choices = choices)
+        } else if ("Date" %in% param$class && identical("Sys.Date()", param$expr)) {
+          useSelectControl <- function(current) {
+            (is.null(current) || identical(current, "default"))
+          }
+          uidefaults[[param$name]] <<- "default"
+          choices <- list()
+          choices[[paste0("today (",param$value,")")]] <- "default"
+          choices[["Use a custom date"]] <- "custom"
+          selectControl <- shiny::selectInput(inputId = inputId,
+                                              label = label,
+                                              choices = choices)
+        }
+      }
+      
+      output[[paste0("ui_", param$name)]] <- shiny::renderUI({
+        if (useSelectControl(input[[param$name]])) {
+          selectControl
+        } else {
+          inputControl
+        }
+      })
+    }
+
+    lapply(configurable, function(param) {
+      param.ui(param)
+    })
+    
+    values <- shiny::eventReactive(input$save, {
+      values <- list()
+      lapply(configurable, function(param) {
+        default <- uidefaults[[param$name]]
+        value <- input[[param$name]]
+        if (!identical(value, default)) {
+          values[[param$name]] <<- params_get_converter(param)(value)
+        }
+      })
+      ## This set of published values is the raw set that came from the user.
+      ## It does not include those values that cannot be configured or are
+      ## left to use the default.
+      values
+    })
+
+    shiny::observe({
+      shiny::stopApp(values())
+    })
+
+    shiny::observeEvent(input$cancel, {
+      shiny::stopApp(NULL)
+    })
+  }
+
+  contents <- shiny::tagList(
+      shiny::fluidRow(shiny::column(12,shiny::tags$h1("Configure Report Parameters"))),
```

Comment:
The problem is that param$value can be NULL (which is why we have a special
param$class rather than using class(param$value). We may want to talk one
more time about whether allowing NULL is necessary/desirable)
On Tue, Jul 21, 2015 at 8:11 AM Aron Atkins <notifications@github.com>
wrote:

> In R/params.R
> <https://github.com/rstudio/rmarkdown/pull/473#discussion_r35094937>:
>
> > +      ## It does not include those values that cannot be configured or are
> > +      ## left to use the default.
> > +      values
> > +    })
> > +
> > +    shiny::observe({
> > +      shiny::stopApp(values())
> > +    })
> > +
> > +    shiny::observeEvent(input$cancel, {
> > +      shiny::stopApp(NULL)
> > +    })
> > +  }
> > +
> > +  contents <- shiny::tagList(
> > +      shiny::fluidRow(shiny::column(12,shiny::tags$h1("Configure Report Parameters"))),
>
> Stepped down this heading and the "cannot be configured" heading below.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rstudio/rmarkdown/pull/473/files#r35094937>.
>